BSP/LED/led.c

```diff
@@ -0,0 +1,86 @@
+#include "led.h"
+#include <stm32l4xx_hal.h>
+
+#define NB_LED_PER_EDGE (9)
+
+typedef struct {
+    GPIO_TypeDef *const port;
+    uint32_t pin;
+} bspIO_s;
+
+/**
+ * @brief list of ANODE LED IOs from LED_A to LED_I
+ */
+static const bspIO_s anodeLedIOtable[NB_LED_PER_EDGE]={
+    {GPIOD, GPIO_PIN_14},
+    {GPIOD, GPIO_PIN_12},
+    {GPIOD, GPIO_PIN_10},
+    {GPIOB, GPIO_PIN_14},
+    {GPIOB, GPIO_PIN_12},
+    {GPIOE, GPIO_PIN_15},
+    {GPIOE, GPIO_PIN_13},
+    {GPIOE, GPIO_PIN_11},
+    {GPIOE, GPIO_PIN_9}
+};
+
+
+/**
+ * @brief list of CATHODE LED IOs from LED_1 to LED_9
+ */
+static const bspIO_s cathodeLedIOtable[NB_LED_PER_EDGE]={
+    {GPIOE, GPIO_PIN_7},
+    {GPIOH, GPIO_PIN_1},
+    {GPIOC, GPIO_PIN_15},
+    {GPIOC, GPIO_PIN_13},
+    {GPIOE, GPIO_PIN_5},
+    {GPIOE, GPIO_PIN_3},
+    {GPIOB, GPIO_PIN_9},
+    {GPIOB, GPIO_PIN_7},
+    {GPIOB, GPIO_PIN_5}
+};
+
+
+void led_init(void)
+{
+    __HAL_RCC_GPIOB_CLK_ENABLE();
+    __HAL_RCC_GPIOC_CLK_ENABLE();
+    __HAL_RCC_GPIOD_CLK_ENABLE();
+    __HAL_RCC_GPIOE_CLK_ENABLE();
+    __HAL_RCC_GPIOH_CLK_ENABLE();
+
+    GPIO_InitTypeDef GPIO_InitStruct;
+    GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_PP;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+    GPIO_InitStruct.Pull = GPIO_NOPULL;
+
+    // cathode needs to be set high before anode to avoid anyblinking during powerup
```

Comment:
Ok for the comment, but I dont see why you have two loops, looks like you can merge them.
Are you missing a Delay?

Reply:
every cathode connected to every anode. It's a 9x9 matrix not 9 lines. So all K must be set before any A I think